File: rust/src/winenv/api/usermode/psapi.rs
use crate::winenv::api::ApiHandler;
use crate::binemu::BinaryEmulator;
use crate::errors::Result;

pub struct PsapiHandler;

impl PsapiHandler {
    pub fn new() -> Self {
        Self
    }

    pub fn enum_processes(&self) -> Vec<u32> {
        vec![4, 888, 1337]
    }

    pub fn enum_process_modules(&self, _process: u32) -> Vec<u64> {
        vec![0x400000, 0x70000000]
    }

    pub fn get_module_base_name(&self, module: u64) -> String {
        match module {
            0x400000 => "sample.exe".to_string(),
            0x70000000 => "ntdll.dll".to_string(),
            _ => "unknown.dll".to_string(),
        }
    }

    pub fn get_module_information(&self, module: u64) -> (u64, u32, u64) {
        (module, 0x1000, module + 0x100)
    }
}

impl Default for PsapiHandler {
    fn default() -> Self {
        Self::new()
    }
}

impl ApiHandler for PsapiHandler {
    fn call(&mut self, _emu: &mut dyn BinaryEmulator, name: &str, args: &[u64]) -> Result<u64> {
        match name {
            "EnumProcesses" => {
                let lpid_process = args[0];
                let cb = args[1] as usize;
                let lpcb_needed = args[2];

                let processes = self.enum_processes();

                if lpcb_needed != 0 {
                    let needed = (processes.len() * 4) as u32;
                    emu.mem_write(lpcb_needed, &needed.to_le_bytes())?;
                }

                if lpid_process == 0 || cb < 4 {
                    return Ok(1);
                }

                let count = (cb / 4).min(processes.len());
                let mut cursor = lpid_process;
                for process in processes.iter().take(count) {
                    emu.mem_write(cursor, &process.to_le_bytes())?;
                    cursor += 4;
                }

                Ok(1)
            },
            "EnumProcessModules" => {
                let _h_process = args[0];
                let lph_module = args[1];
                let cb = args[2] as usize;
                let lpcb_needed = args[3];

                let modules = self.enum_process_modules(0);

                if lpcb_needed != 0 {
                    let needed = (modules.len() * 8) as u32;
                    emu.mem_write(lpcb_needed, &needed.to_le_bytes())?;
                }

                if lph_module == 0 || cb < 8 {
                    return Ok(1);
                }

                let count = (cb / 8).min(modules.len());
                let mut cursor = lph_module;
                for module in modules.iter().take(count) {
                    emu.mem_write(cursor, &module.to_le_bytes())?;
                    cursor += 8;
                }

                Ok(1)
            },
            "GetModuleBaseName" | "GetModuleBaseNameA" | "GetModuleBaseNameW" => {
                let _h_process = args[0];
                let h_module = args[1];
                let lp_base_name = args[2];
                let n_size = args[3] as usize;

                if lp_base_name == 0 || n_size == 0 {
                    return Ok(0);
                }

                let module_name = self.get_module_base_name(h_module);
                let truncated = if module_name.len() > n_size - 1 {
                    &module_name[..n_size - 1]
                } else {
                    &module_name
                };

                let output = format!("{}\0", truncated);
                emu.mem_write(lp_base_name, output.as_bytes())?;

                Ok(truncated.len() as u64)
            },
            "GetModuleFileNameEx" | "GetModuleFileNameExA" | "GetModuleFileNameExW" => {
                let _h_process = args[0];
                let h_module = args[1];
                let lp_filename = args[2];
                let n_size = args[3] as usize;

                if lp_filename == 0 || n_size == 0 {
                    return Ok(0);
                }

                let module_name = self.get_module_base_name(h_module);
                let truncated = if module_name.len() > n_size - 1 {
                    &module_name[..n_size - 1]
                } else {
                    &module_name
                };

                let output = format!("{}\0", truncated);
                emu.mem_write(lp_filename, output.as_bytes())?;

                Ok(truncated.len() as u64)
            },
            _ => Ok(0),
        }
    }

    fn get_name(&self) -> &str {
        "Psapi"
    }
}

File: secpp/winenv/api/usermode/psapi.cpp
// psapi.cpp  psapi.dll handler (real implementations)
#include "psapi.h"
#include <cstring>
#include <cstdint>
#include <string>
#include <vector>
#include <algorithm>
#include "windows/winemu.h"
#include "windows/objman.h"
#include "windows/common.h"

using namespace speakeasy;

namespace speakeasy { namespace api {

static inline int ptr_sz(void* e) { return (be(e)->get_arch() == speakeasy::arch::ARCH_AMD64) ? 8 : 4; }

Psapi::Psapi(void* emu) : ApiHandler(emu) {
    INIT_API_TABLE(Psapi)
    REG(Psapi, EnumProcesses, 3)        REG(Psapi, EnumProcessModules, 4)
    REG(Psapi, GetModuleBaseName, 4)    REG(Psapi, GetModuleBaseNameA, 4)
    REG(Psapi, GetModuleBaseNameW, 4)   REG(Psapi, GetModuleFileNameEx, 4)
    REG(Psapi, GetModuleFileNameExA, 4) REG(Psapi, GetModuleFileNameExW, 4)
    END_API_TABLE
}

// Helper: get module bases for a process
static std::vector<uint64_t> get_process_module_bases(void* e, std::shared_ptr<Process> proc) {
    std::vector<uint64_t> bases;
    if (!proc) return bases;

    auto module = proc->pe;
    if (module) {
        bases.push_back(module->base);
        return bases;
    } 

    std::string p = proc->path;
    if (!p.empty()) {
        size_t slash = p.rfind('\\');
        std::string mod_name;
        if (slash != std::string::npos) mod_name = p.substr(slash + 1);
        else mod_name = p;
        size_t dot = mod_name.rfind('.');
        if (dot != std::string::npos) mod_name = mod_name.substr(0, dot);
        auto mod = we(e)->get_mod_by_name(mod_name);
        if (mod) {
            bases.push_back(mod->base);
            return bases;
        }
    }

    uint64_t process_base = proc->base;
    if (process_base) {
        bases.push_back(process_base);
    }

    return bases;
}

// Helper: get module base name
static std::string get_module_base_name(void* e, std::shared_ptr<Process> proc, uint64_t hModule) {
    if (hModule) {
        auto mod = we(e)->get_mod_from_addr(hModule);
        if (mod) {
            std::string ep = mod->emu_path;
            size_t slash = ep.rfind('\\');
            if (slash != std::string::npos) return ep.substr(slash + 1);
            return ep;
        }
    }
    std::string p = proc->path;
    size_t slash = p.rfind('\\');
    if (slash != std::string::npos) return p.substr(slash + 1);
    return p;
}

// Helper: get module file name (full path)
static std::string get_module_file_name(void* e, std::shared_ptr<Process> proc, uint64_t hModule) {
    if (hModule) {
        auto mod = we(e)->get_mod_from_addr(hModule);
        if (mod) return mod->emu_path;
    }
    return proc->path;
}

// 
//  EnumProcesses  enumerate running processes
// 
uint64_t Psapi::EnumProcesses(void* e, ArgList& a, void* ctx) {
    uint64_t lpidProcess = a[0];
    uint64_t cb = a[1];
    uint64_t lpcbNeeded = a[2];

    auto& processes = we(e)->get_processes();
    uint32_t count = static_cast<uint32_t>(processes.size());

    // Write needed size
    if (lpcbNeeded) {
        uint32_t needed = count * 4;
        std::vector<uint8_t> nb(4, 0);
        speakeasy::write_le(nb, 0, static_cast<uint64_t>(needed), 4);
        we(e)->mem_write(lpcbNeeded, nb);
    }

    if (!lpidProcess || cb < 4) return 1;

    uint32_t max_write = static_cast<uint32_t>(std::min(cb / 4, static_cast<uint64_t>(count)));
    uint64_t cursor = lpidProcess;
    for (uint32_t i = 0; i < max_write; i++) {
        auto proc = processes[i];
        uint32_t pid = proc ? static_cast<uint32_t>(proc->get_pid()) : 0;
        std::vector<uint8_t> pid_buf(4, 0);
        speakeasy::write_le(pid_buf, 0, static_cast<uint64_t>(pid), 4);
        we(e)->mem_write(cursor, pid_buf);
        cursor += 4;
    }

    return 1;
}

// 
//  EnumProcessModules  enumerate modules in a process
// 
uint64_t Psapi::EnumProcessModules(void* e, ArgList& a, void* ctx) {
    uint64_t hProcess = a[0];
    uint64_t lphModule = a[1];
    uint64_t cb = a[2];
    uint64_t lpcbNeeded = a[3];

    auto proc = std::dynamic_pointer_cast<Process>(we(e)->get_object_from_handle(hProcess));
    if (!proc) return 0;

    auto bases = get_process_module_bases(e, proc);
    int ps = ptr_sz(e);

    // Write needed size
    if (lpcbNeeded) {
        uint32_t needed = static_cast<uint32_t>(bases.size() * ps);
        std::vector<uint8_t> nb(4, 0);
        speakeasy::write_le(nb, 0, static_cast<uint64_t>(needed), 4);
        we(e)->mem_write(lpcbNeeded, nb);
    }

    if (!lphModule || cb < static_cast<uint64_t>(ps)) return 1;

    uint32_t max_write = static_cast<uint32_t>(std::min(cb / ps, static_cast<uint64_t>(bases.size())));
    uint64_t cursor = lphModule;
    for (uint32_t i = 0; i < max_write; i++) {
        std::vector<uint8_t> mb(ps, 0);
        speakeasy::write_le(mb, 0, bases[i], ps);
        we(e)->mem_write(cursor, mb);
        cursor += ps;
    }

    return 1;
}

// 
//  GetModuleBaseName / GetModuleBaseNameA / GetModuleBaseNameW
// 
uint64_t Psapi::GetModuleBaseName(void* e, ArgList& a, void* ctx) {
    uint64_t hProcess = a[0];
    uint64_t hModule = a[1];
    uint64_t lpBaseName = a[2];
    uint64_t nSize = a[3];

    if (!lpBaseName || nSize == 0) return 0;

    auto proc = std::dynamic_pointer_cast<Process>(we(e)->get_object_from_handle(hProcess));
    if (!proc) return 0;

    std::string name = get_module_base_name(e, proc, hModule);
    if (name.empty()) return 0;

    if (name.size() >= nSize) name = name.substr(0, nSize - 1);

    be(e)->write_mem_string(name, lpBaseName, 1);
    return static_cast<uint64_t>(name.size());
}

uint64_t Psapi::GetModuleBaseNameA(void* e, ArgList& a, void* ctx) {
    return GetModuleBaseName(e, a, ctx);
}

uint64_t Psapi::GetModuleBaseNameW(void* e, ArgList& a, void* ctx) {
    uint64_t hProcess = a[0];
    uint64_t hModule = a[1];
    uint64_t lpBaseName = a[2];
    uint64_t nSize = a[3];

    if (!lpBaseName || nSize == 0) return 0;

    auto proc = std::dynamic_pointer_cast<Process>(we(e)->get_object_from_handle(hProcess));
    if (!proc) return 0;

    std::string name = get_module_base_name(e, proc, hModule);
    if (name.empty()) return 0;

    if (name.size() >= nSize) name = name.substr(0, nSize - 1);

    be(e)->write_mem_string(name, lpBaseName, 2);
    return static_cast<uint64_t>(name.size());
}

// 
//  GetModuleFileNameEx / GetModuleFileNameExA / GetModuleFileNameExW
// 
uint64_t Psapi::GetModuleFileNameEx(void* e, ArgList& a, void* ctx) {
    uint64_t hProcess = a[0];
    uint64_t hModule = a[1];
    uint64_t lpFilename = a[2];
    uint64_t nSize = a[3];

    if (!lpFilename || nSize == 0) return 0;

    auto proc = std::dynamic_pointer_cast<Process>(we(e)->get_object_from_handle(hProcess));
    if (!proc) return 0;

    std::string name = get_module_file_name(e, proc, hModule);
    if (name.empty()) return 0;

    if (name.size() >= nSize) name = name.substr(0, nSize - 1);

    be(e)->write_mem_string(name, lpFilename, 1);
    return static_cast<uint64_t>(name.size());
}

uint64_t Psapi::GetModuleFileNameExA(void* e, ArgList& a, void* ctx) {
    return GetModuleFileNameEx(e, a, ctx);
}

uint64_t Psapi::GetModuleFileNameExW(void* e, ArgList& a, void* ctx) {
    uint64_t hProcess = a[0];
    uint64_t hModule = a[1];
    uint64_t lpFilename = a[2];
    uint64_t nSize = a[3];

    if (!lpFilename || nSize == 0) return 0;

    auto proc = std::dynamic_pointer_cast<Process>(we(e)->get_object_from_handle(hProcess));
    if (!proc) return 0;

    std::string name = get_module_file_name(e, proc, hModule);
    if (name.empty()) return 0;

    if (name.size() >= nSize) name = name.substr(0, nSize - 1);

    be(e)->write_mem_string(name, lpFilename, 2);
    return static_cast<uint64_t>(name.size());
}

}} // namespaces
